### PR TITLE
introducing BigPixel for phase-2 InnerTracker  in T39 geometry

### PR DESCRIPTION
**Pull request by emiglior**
#### PR description:

This PR is a follow-up of the draft [PR#42454](https://github.com/cms-sw/cmssw/pull/42454) (closed in Jan 2024).

The PR introduces the so-called BigPixels (elongated pixels between two readout chips) for the phase-2 InnerTracker following the implementation proposed in:

https://indico.cern.ch/event/1382902/contributions/5816585/attachments/2805475/4895325/Phase-2_IT_BigPix_TrkDPG.pdf

- **Description.** BigPixels are introduced in the file pixelStructureTopology.xml or by using the TrackerAdditionalParametersPerDet rcd for geometries loaded from DB (none yet). Therefore, the PR restores the changes of the infrastructure needed for using the TrackerAdditionalParametersPerDet rcd, introduced for the bricked pixels and then removed with [PR#40443](https://github.com/cms-sw/cmssw/pull/40443) but without reintroducing the functions related to bricked pixels.
**NOTE (similar to [PR#34120](https://github.com/cms-sw/cmssw/pull/34120)): the record PTrackerAdditionalParametersPerDetRcd is consumed by the TrackerGeomBuilderFromGeometricDet, which is the same for Runs 2&3 and Phase 2. Therefore, this PR also modifies Runs 2&3, by including the new record.** 

- **Tracker geometry.** The Tracker phase-2 test geometry in which BigPixels are introduced is T39, thus skipping versions T36-T37-T38 which are currently under development by @Raffaella07.

- **Implementation.** Phase-2 pixels are now described by the class RectangularPhase2PixelTopology inheriting from PixelTopology. Code for Tracker local reco was adapted to work both the phase-1 pixels (RectangularPixelTopology) and with phase-2 pixels (RectangularPhase2PixelTopology). The function siPixelUtils::generic_position_formula(), used in PixelCPEGeneric, was modified in order to be used both for phase-1 and phase-2 BigPixels. A side-effect was to add to the abstract class PixelTopology the methods getPixelFractionInX() and getPixelFractionInY(). @fabiocos @martinamalberti: the implementation of these fcn in ProxyMTDTopology is just a placeholder to compile the code. Please let us know what the actual code should be.

#### PR validation:
    
compilation ok after git cms-checkdeps -a -A
scram b code-checks, code-format
scram b runtests
runTheMatrix.py -l limited -i all --ibeos

No changes are expected on phase-1 (set of tracking validation plots on a single muon sample on phase-1 geometry)
 http://personalpages.to.infn.it/~migliore/BigPixels/SingleMuPt100-phase1/plots-20240323/

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

This PR is not a backport and no backport needed.

Specific w/f for test on T39: 30034.0

@gbardell @langufo @bartosik-hep 
@mmusich @sroychow @mdelcourt FYI

### CODE REVIEW

**Comment by mmusich** [comment]
@cmsbuild, please test

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 1]
```diff
@@ -244,6 +244,7 @@ def checkHLTfor43774(process):
 
     return process
     
+
 # CMSSW version specific customizations
 def customizeHLTforCMSSW(process, menuType="GRun"):
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py:249]: Please remove this irrelevant change and use the original file (one signature less needed)

Comment (mmusich) [reply HLTrigger/Configuration/python/customizeHLTforCMSSW.py:249]: https://github.com/cms-sw/cmssw/pull/42454 required a change in `customizeHLTforCMSSW`, see https://github.com/cms-sw/cmssw/pull/42454/files#r1285949345. Is this not needed also here?

Comment (mmusich) [reply HLTrigger/Configuration/python/customizeHLTforCMSSW.py:249]: and indeed this PR is failing integration tests, see. e.g [log file](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3f5fe0/38502/addOnTests/hlt_mc_GRun/step2.log).

@emiglior please take note.

**Comment by mmusich** [comment]
In addition to the discussion above at https://github.com/cms-sw/cmssw/pull/44576#discussion_r1544282401 (existing  HLT menus in release need to be patched in order to support the new record), the PR is also failing the [unit tests](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3f5fe0/38502/unitTests/failed.html):
   * [test2021Geometry](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3f5fe0/38502/unitTests/src/Configuration/Geometry/test/test2021Geometry/testing.log)
   * [test2026Geometry](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3f5fe0/38502/unitTests/src/Configuration/Geometry/test/test2026Geometry/testing.log)

**Comment by emiglior** [comment]
@Martin-Grunewald @mmuich thanks for the feedback - please wait to restart testing - still fixin the error in the test2024 and test2026 units tests

**Review by mmusich — COMMENTED** [review]
you might as well clean-up unused code.

**Comment by mmusich** [comment]
before testing again, would it be possible to squash the commits (in particular to get rid of the merges of PRs to the branch)?

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
@cmsbuild, please abort

**Comment by mmusich** [comment]
test parameters:
   * workflow_opts = -w upgrade
   * workflows = 30034.0

**Comment by mmusich** [comment]
@cmsbuild, please test

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 2]
```diff
@@ -244,6 +244,13 @@ def checkHLTfor43774(process):
 
     return process
     
+def customizeHLTFor44576(process):
```
Comment (mmusich) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py:247]:
```suggestion
def customizeHLTfor44576(process):
```

@emiglior @bardellig please test the branch before submitting more updates.

**Comment by emiglior** [comment]
Could somebody kindly start the tests?

**Comment by mmusich** [comment]
> Could somebody kindly start the tests?

was it tested privately?

**Comment by civanch** [comment]
please test

**Comment by emiglior** [comment]
> > Could somebody kindly start the tests?
> 
> was it tested privately?

at least for the issues related to the typo in customizeHLTfor44576() but some test failed because of problems in accessing the input dataset

**Comment by mmusich** [comment]
> at least for the issues related to the typo in customizeHLTfor44576() but some test failed because of problems in accessing the input dataset

in the previous round of tests there were still unit test problems:

```
---> test test2021Geometry had ERRORS
---> test test2026Geometry had ERRORS
```

are they solved in the last push?

**Comment by mmusich** [comment]
For HLT this is a somewhat simple change and therefore can be signed right away.
OTOH there might be need to adjust the signature if other coordination areas have comments. It would be better if at lease @cms-sw/reconstruction-l2 and @cms-sw/geometry-l2 can have a look / sign this first.

**Review by perrotta — COMMENTED** [review]
From AlCa only a few minor comments on the code, that you can take into account if you plan to update it further

**Comment by aloeliger** [comment]
L1T would seem to be somewhat incidental to this PR, just removed unused lines of code, so I think we're fine with this.

**Comment by aloeliger** [comment]
+l1

**Comment by mmusich** [comment]
code-checks

**Comment by mmusich** [comment]
@emiglior @langufo @bardellig, this branch has now conflicts that need to be resolved.

**Comment by emiglior** [comment]
> @emiglior @langufo @bardellig, this branch has now conflicts that need to be resolved.

yes, we know... we are looking into this

**Comment by emiglior** [comment]
branch updated to fix conflicts after PR#44765 was merged into master

w/f for test on T39/2026D114: 31234.0

**Comment by mmusich** [comment]
test parameters:
   * workflow = 31234.0

**Comment by bsunanda** [comment]
+geometry

**Comment by aloeliger** [comment]
+l1

**Comment by AdrianoDee** [comment]
+pdmv

**Comment by perrotta** [comment]
+1

**Comment by jfernan2** [comment]
+1

**Comment by bsunanda** [comment]
@srimanob Please sign this

**Comment by srimanob** [comment]
+Upgrade

**Comment by tjavaid** [comment]
+1

**Comment by mmusich** [comment]
@cmsbuild please test
  * refreshing two weeks old tests.

**Comment by mmusich** [comment]
@cmsbuild, please test
   * failures are transient

**Comment by emiglior** [comment]
In case of phase-2 w/f, we changed the binning of the th2 digi-map
DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cff.py
to better spot the presence of big pixels

**Comment by mmusich** [comment]
> In case of phase-2 w/f, we changed the binning of the th2 digi-map

this explains this.

>  DQMHistoSizes: changed ( 23234.0,... ): 36994.805 KiB TrackerPhase2ITDigi/DigiMonitor

what about the other failures?

**Comment by emiglior** [comment]
Still on ph-2 w/f, IIUC the failures are actually fired by tiny differences (but let me know if I am wrong). We observed something similar during the development of the code. 
We ascribed the effect to the re-coding of the functions

localX(), localY(), pixel() 

in Geometry/TrackerGeometryBuilder/src/RectangularPixelPhase2Topology.cc  (e.g. different rounding-off's)

**Comment by emiglior** [comment]
hi, any news on the missing signatures?

**Comment by rappoccio** [comment]
+1

**Comment by mmusich** [comment]
@emiglior looks like this PR causes unit tests failures, see e.g. [here](https://cmssdt.cern.ch/SDT/cgi-bin/logreader/el8_amd64_gcc12/CMSSW_14_1_X_2024-05-21-2300/unitTestLogs/DQM/SiStripCommon#/37)

**Comment by mmusich** [comment]
> looks like this PR causes unit tests failures.

https://github.com/cms-sw/cmssw/pull/45013 should fix.

**Comment by dan131riley** [comment]
DBG builds are broken, see for instance https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el8_amd64_gcc12/CMSSW_14_1_DBG_X_2024-05-23-2300/Geometry/TrackerGeometryBuilder
```
>> Compiling  src/Geometry/TrackerGeometryBuilder/src/trackerHierarchy.cc
/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DDD4HEP_USE_GEANT4_UNITS=1 -DCMSSW_GIT_HASH='CMSSW_14_1_DBG_X_2024-05-23-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_1_DBG_X_2024-05-23-2300' -Isrc -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/dd4hep/v01-27-02-02d087b69a43429d82df59691bcd25e1/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/pcre/8.43-e34796d17981e9b6d174328c69446455/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/boost/1.80.0-941b136a4a3be6f8bc1e903d36ddc172/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/clhep/2.4.7.1-8e40efd27b7394c1fa4e9c7e432d85cd/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/gsl/2.6-5e2ce72ea2977ff21a2344bbb52daf5c/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/lcg/root/6.30.07-f3322c77db1c59847b28fde88ff7218c/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/tbb/v2021.9.0-a7089dd5ec356e9a0bc222e109b15cef/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/cms/vdt/0.4.3-f094bee80112624813c07f9336e08d7d/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/zlib/1.2.11-1a082fc322b0051b504cc023f21df178/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-3ca740c03e68b1a067f3ed0679234a78/include/eigen3 -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/fmt/8.0.1-258b4791803c34b7e98cf43693e54d87/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/OpenBLAS/0.3.15-c877ab57fa7b04ce290093588c6c5717/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/tinyxml2/6.2.0-88fe0ec301baf763fa3c485e5b67ed91/include -O2 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++17 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v2 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -g -O3 -DEDM_ML_DEBUG -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/Geometry/TrackerGeometryBuilder/src/GeometryTrackerGeometryBuilder/trackerHierarchy.cc.d src/Geometry/TrackerGeometryBuilder/src/trackerHierarchy.cc -o tmp/el8_amd64_gcc12/src/Geometry/TrackerGeometryBuilder/src/GeometryTrackerGeometryBuilder/trackerHierarchy.cc.o
src/Geometry/TrackerGeometryBuilder/src/RectangularPixelPhase2Topology.cc: In member function 'virtual std::pair<float, float> RectangularPixelPhase2Topology::pixel(const LocalPoint&) const':
  src/Geometry/TrackerGeometryBuilder/src/RectangularPixelPhase2Topology.cc:86:104: error: 'numROC' was not declared in this scope
    86 |                                                << iybin << " " << fractionY << " " << iybin0 << " " << numROC;
      |                                                                                                        ^~~~~~
  src/Geometry/TrackerGeometryBuilder/src/RectangularPixelPhase2Topology.cc:95:104: error: 'numROC' was not declared in this scope
    95 |                                                << iybin << " " << fractionY << " " << iybin0 << " " << numROC;
      |                                                                                                        ^~~~~~
  src/Geometry/TrackerGeometryBuilder/src/RectangularPixelPhase2Topology.cc:133:16: error: 'm_ROW_PER_ROC' was not declared in this scope; did you mean 'm_ROWS_PER_ROC'?
   133 |   if (ixbin0 > m_ROW_PER_ROC || ixbin0 < 0)  //  ixbin < 0 outside range
      |                ^~~~~~~~~~~~~
      |                m_ROWS_PER_ROC
  src/Geometry/TrackerGeometryBuilder/src/RectangularPixelPhase2Topology.cc:143:30: error: 'm_ROW_PER_ROC' was not declared in this scope; did you mean 'm_ROWS_PER_ROC'?
   143 |   if (mpX < 0. || mpX >= 2 * m_ROW_PER_ROC) {
      |                              ^~~~~~~~~~~~~
      |                              m_ROWS_PER_ROC
  gmake: *** [tmp/el8_amd64_gcc12/src/Geometry/TrackerGeometryBuilder/src/GeometryTrackerGeometryBuilder/RectangularPixelPhase2Topology.cc.o] Error 1
```

**Comment by mmusich** [comment]
> DBG builds are broken, see for instance https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el8_amd64_gcc12/CMSSW_14_1_DBG_X_2024-05-23-2300/Geometry/TrackerGeometryBuilder

https://github.com/cms-sw/cmssw/pull/45038 ?

**Pull request merged**